File: app/src/main/java/com/antyzero/smoksmog/time/PolishCountdown.java
package com.antyzero.smoksmog.time;

public class PolishCountdown implements Countdown {

    @Override
    public String get( int givenSeconds ) {

        int seconds = givenSeconds & 60;
        int minutes = ( givenSeconds / 60 ) % 60;
        int hours = ( givenSeconds / 60 ) / 60;

        if ( minutes == 0 ) {
            return secondsAgo( seconds );
        } else if ( hours == 0 ) {
            return minutesAgo( minutes );
        }

        return hoursAgo( hours );
    }

    private String hoursAgo( int hours ) {
        return ago( hours, "godzinę", "godziny", "godzin" );
    }

    private String minutesAgo( int minutes ) {
        return ago( minutes, "minutę", "minuty", "minut" );
    }

    private String secondsAgo( int seconds ) {
        return ago( seconds, "sekundę", "sekundy", "sekund" );
    }

    private String ago( int amount, String single, String some, String many ) {
        if ( amount == 1 ) {
            return amount + " " + single;
        } else if ( endsWithTwoToFour( amount ) ) {
            return amount + " " + some;
        } else {
            return amount + " " + many;
        }
    }

    private boolean endsWithTwoToFour( int seconds ) {
        int modulo = seconds % 10;
        if ( seconds >= 10 && seconds < 20 ) {
            return false;
        } else if ( modulo >= 2 && modulo <= 4 ) {
            return true;
        }
        return false;
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/time/PolishCountdown.kt
package com.antyzero.smoksmog.time

class PolishCountdown : Countdown {

    override fun get(givenSeconds: Int): String {

        val seconds = givenSeconds and 60
        val minutes = givenSeconds / 60 % 60
        val hours = givenSeconds / 60 / 60

        if (minutes == 0) {
            return secondsAgo(seconds)
        } else if (hours == 0) {
            return minutesAgo(minutes)
        }

        return hoursAgo(hours)
    }

    private fun hoursAgo(hours: Int): String {
        return ago(hours, "godzinę", "godziny", "godzin")
    }

    private fun minutesAgo(minutes: Int): String {
        return ago(minutes, "minutę", "minuty", "minut")
    }

    private fun secondsAgo(seconds: Int): String {
        return ago(seconds, "sekundę", "sekundy", "sekund")
    }

    private fun ago(amount: Int, single: String, some: String, many: String): String {
        if (amount == 1) {
            return amount.toString() + " " + single
        } else if (endsWithTwoToFour(amount)) {
            return amount.toString() + " " + some
        } else {
            return amount.toString() + " " + many
        }
    }

    private fun endsWithTwoToFour(seconds: Int): Boolean {
        val modulo = seconds % 10
        if (seconds >= 10 && seconds < 20) {
            return false
        } else if (modulo >= 2 && modulo <= 4) {
            return true
        }
        return false
    }
}
